src/statmodels.jl

```diff
@@ -197,7 +197,7 @@ Coefficient of determination (R-squared).
 For a linear model, the R² is defined as ``ESS/TSS``, with ``ESS`` the explained sum of squares
 and ``TSS`` the total sum of squares.
 """
-r2(obj::StatisticalModel) = mss(obj) / deviance(obj)
+r2(obj::StatisticalModel) = mss(obj) / sum(abs2, response(obj) .- meanresponse(obj))
```

Comment:
Better do `μ = meanresponse(obj); sum(x -> abs2(x - μ), response(obj))` to avoid allocating a copy. Maybe turn this into a long-form function for readability.

@Nosferican Is this formula OK for your package?

Reply:
> avoid allocating a copy

I thought that `sum` didn't allocate transform scalars into vectors with the dot syntax.

> we'd better deprecate this function and leave packages to implement it for linear models

Do you mean removing the function and leaving a warning or leaving the function and adding a warning?

Reply:
Consider this script,
```
using BenchmarkTools
a(x, y) = sum(abs2, x - y)
b(x, y) = sum(abs2, x .- y)
c(x, y) = sum(xy -> abs2(xy[1] - xy[2]), zip(x, y))
d(x, y) = sum(abs2(a - b) for (a, b) ∈ zip(x, y))
using Random: seed!
seed!(0)
const x = rand(1000);
const y = rand(1000);
@btime a($x, $y)
@btime b($x, $y)
@btime c($x, $y)
@btime d($x, $y)
  703.416 ns (1 allocation: 7.94 KiB)
  659.205 ns (1 allocation: 7.94 KiB)
  823.250 ns (1 allocation: 32 bytes)
  969.684 ns (2 allocations: 48 bytes)
```
It seems to be trivially faster, but it certainly allocates as opposed to `c` or `d`.

If the `R²` matches Stata, it will match mine. Some models will still need to define the method for some panel estimators, but this seems good.

If the statsmodels functions are being moved to StatsModels it might be best to make the decision there (better to trim there, than to have lost useful code then),

Reply:
@lbittarello:
> I thought that sum didn't allocate transform scalars into vectors with the dot syntax.

As @Nosferican shows, it allocates, and it will probably stay that way in the future. Maybe another syntax will be introduced to do that.

> Do you mean removing the function and leaving a warning or leaving the function and adding a warning?

I mean print a warning using `Base.depwarn(..., :r2)`, saying that packages should define their custom method. Then we'll turn the warning into an error just like for other function stubs in the file.

@Nosferican:
> If the R² matches Stata, it will match mine. Some models will still need to define the method for some panel estimators, but this seems good.

OK, cool.

> If the statsmodels functions are being moved to StatsModels it might be best to make the decision there (better to trim there, than to have lost useful code then),

It's not clear yet what will happen and when, and given how small this function is that's really not an issue.

Reply:
I've added a deprecation warning. Let me know if it's okay.